replace FAILED deployments with `helm upgrade --install --force`

bacongobbler opened the pull request:
When using `helm upgrade --install`, if the first release fails, Helm will respond with an error saying that it cannot upgrade from an unknown state.

With this feature, `helm upgrade --install --force` automates the same process as `helm delete && helm install --replace`. It will mark the previously FAILED release as DELETED, delete any existing resources inside Kubernetes, then replace it as if it was a fresh install. I did not want to make this the default behaviour of `helm upgrade --install` because this is a destructive operation that deletes resources in Kubernetes, and the operator should opt into and accept this behaviour.

closes #3353 
refs discussion in #3437

Commit:
replace FAILED deployments with `helm upgrade --install --force`

When using `helm upgrade --install`, if the first release fails, Helm will respond with an error saying that it cannot upgrade from an unknown state.

With this feature, `helm upgrade --install --force` automates the same process as `helm delete && helm install --replace`. It will mark the previous release as DELETED, delete any existing resources inside Kubernetes, then replace it as if it was a fresh install. It will then mark the FAILED release as SUPERSEDED.

bacongobbler commented:
Forward note: I'm a little iffy on marking the previous release as SUPERSEDED.

helgi commented:
Can this be the default behaviour if there is only 1 release prior and it failed? Seems to be the most common case so far. 

Beyond that, I agree with it not being the default behaviour.

thomastaylor312 reviewed (commented):
One small comment for discussion, otherwise I tested this and it is good to go

bacongobbler commented on pkg/tiller/release_update_test.go:
```diff
@@ -239,6 +239,50 @@ func TestUpdateReleaseFailure(t *testing.T) {
+
+	compareStoredAndReturnedRelease(t, *rs, *res)
+
+	edesc := "Upgrade complete"
```
clearer variable names are always great.

bacongobbler commented on pkg/tiller/release_update_test.go:
```diff
@@ -239,6 +239,50 @@ func TestUpdateReleaseFailure(t *testing.T) {
+		Chart: &chart.Chart{
+			Metadata: &chart.Metadata{Name: "hello"},
+			Templates: []*chart.Template{
+				{Name: "templates/something", Data: []byte("hello: world")},
```
That can be arranged...

bacongobbler commented on pkg/tiller/release_update.go:
```diff
@@ -137,6 +142,97 @@ func (s *ReleaseServer) prepareUpdate(req *services.UpdateReleaseRequest) (*rele
+		}
+	}
+
+	oldRelease.Info.Status.Code = release.Status_SUPERSEDED
```
Hmm good point. I also think it might be good to just keep this in the FAILED state so others cannot roll back to this release, which SUPERSEDED or REPLACED would allow. The more I think about this, the more I would prefer to retain the existing FAILED state.

Related comment: https://github.com/kubernetes/helm/pull/3597#issuecomment-370050089

@adamreese any opinions on this?

thomastaylor312 replied:
I think FAILED would also be a good idea

bacongobbler commented:
> Can this be the default behaviour if there is only 1 release prior and it failed? Seems to be the most common case so far.

That is good feedback, thanks @helgi. I'd be a little concerned about the behaviour being inconsistent for users though. In this case, a user can expect `helm upgrade --install` to "fix" the first failed release, but it will continue to fail on subsequent releases with no feedback on how to fix that. I'd kinda prefer to make it explicitly opt-in as a feature flag, but I'd love to hear any feedback on whether that's important!

helgi commented:
> Perhaps we can decide at a later date if we should do that based on others' feedback in 2.8.2. How does that sound?

Yeah, I considered the inconsistency and cringed as I wrote that message. The scenario I run into a little too often (for comfort) is engineers throwing together a new helm chart and not running it in minikube but rather putting it directly into the dev CI system (copy pasta basically), which leads to failed first deployment a lot of the time. 

The user generally has no idea why it failed and why their future pushes do not fix the issue, and even when they have done the fix before (helm delete --purge) it is forgotten a lot of the time. Basically, first time deploy clumsiness UX issues.

I'm happy with deferring the decision but I did want to bring up the use case

bacongobbler commented on pkg/tiller/release_update.go:
```diff
@@ -137,6 +142,97 @@ func (s *ReleaseServer) prepareUpdate(req *services.UpdateReleaseRequest) (*rele
+		return res, err
+	}
+
+	// pre-upgrade hooks
```
yes

bacongobbler commented:
k, addressed all comments. New stuff since the last round of reviews:

- the old release finishes in state DELETED instead of state SUPERSEDED, modelling what `helm install --replace` would do
- due to the wonky semantics of this feature flag, pre/post delete and install hooks are run, not pre/post upgrade hooks 
- m0ar Star Wars references in the tests for @thomastaylor312's enjoyment


Should be good for another round of reviews

thomastaylor312 reviewed (approved):
This looks good, and your Star Wars reference is golden

mcfedr commented:
Am I write in reading this as `--force` will only ever cause a delete and redeploy when the first deploy has failed? 

so if i have a chart deployed, break it, and then upgrade it, it will be fixed, not recreated?

stealthybox commented:
@bacongobbler I'm also confused regarding this change.
```
helm upgrade --install
```
- if release v1 failed, release v2 will fail because the operation is considered unsafe
- if release v4 succeeded and release v5 failed, release v6 will succeed based off of release v4  ? 
```
helm upgrade --install --force
```
- if release v1 failed, release v2 will succeed because release v1 is deleted first
- if release v4 succeeded and release v5 failed, release v6 will succeed based off of release ...  ?
  does it delete the pre-existing release?
  does it cause downtime?

Is the `--force` flag safe to use all the time, expecting that it only destroys kubernetes resources when the very first release fails?

bacongobbler commented:
`helm update --install` with the `--force` flag automates what one would do to "fix" a failed upgrade. `--force` is just a `helm delete && helm install --replace`, and it only kicks in when the release failed to deploy. It only causes downtime if your application would go into a failed state. `--force` just attempts to fix it. If your application would normally upgrade gracefully, there's no downtime.

bacongobbler commented:
and no, it kicks in *any* time a release fails.

stealthybox commented:
is case 2 accurate?
I'm already lost as to whether helm rolls back release versions on failure

stealthybox commented:
We're still on Helm v2.7.0 because the current upgrade-over-failure behavior appears to be safer for our use case than deleting a deployment.

Our releases usually fail due to hooks, but our hooks are idempotent Jobs, so it's usually pretty safe to upgrade right over them.

If I'm understanding the new behavior correctly, I believe it would be possible for a Deployment to be deleted and for no Pods to be available to serve traffic mid-deployment if a hook failed on the previous deployment.

bacongobbler commented:
To answer case 2, `helm upgrade --install --force` will upgrade as normal, so it'll use v4 to upgrade. However, should v6 fail, that's when the --force flag kicks in and I'm not 100% sure I can recall exactly what happens. It's been a little while so I'll have to look through the code again to answer your question. Feel like looking at it together at KubeCon? :)

stealthybox commented:
:+1: yep, I'm curious
looking forward to it

stealthybox commented:
> If I'm understanding the new behavior correctly, I believe it would be possible for a Deployment to be deleted and for no Pods to be available to serve traffic mid-release if a hook failed on the previous release.

It seems from https://github.com/kubernetes/helm/issues/3208#issuecomment-390340881 that this might be the case when using `upgrade install --force`.
This flag is pretty dangerous.
